### PR TITLE
remove redundant includes in all files

### PR DESCRIPTION
**Pull request by srajangarg**
This was done using a script, so there maybe errors. Will fix.

### CODE REVIEW

**Comment by srajangarg** [comment]
Can't believe there were over 500 instances of header includes that were implicitly being included :stuck_out_tongue:

**Comment by Sumith1896** [comment]
Thank @srajangarg I always intended to do this, never got time :smile:

**Comment by srajangarg** [comment]
Formatting erros had crept into master. Plus somehow [this bug](https://github.com/symengine/symengine/pull/981/files#diff-0bb4632078191d9b93f7d32f38fd6965R347) also crept in. The PR covers both of these issues, apart from the header file removals too.

File: symengine/eval_arb.h [hunk 1]
```diff
@@ -7,8 +7,6 @@
 #ifndef SYMENGINE_EVAL_ARB_H
 #define SYMENGINE_EVAL_ARB_H
 
-#include <symengine/symengine_config.h>
```
Comment (isuruf) [inline symengine/eval_arb.h]: This is definitely needed

File: symengine/basic.h [hunk 2]
```diff
@@ -8,17 +8,13 @@
...
-#include <stdexcept>
-#include <iostream>
 #include <sstream>
 #include <typeinfo>
 #include <map>
 #include <vector>
 #include <set>
 #include <unordered_map>
 #include <cassert>
-#include <ciso646>
```
Comment (isuruf) [inline symengine/basic.h]: This too

File: symengine/basic.h [hunk 3]
```diff
@@ -8,17 +8,13 @@
 #define SYMENGINE_BASIC_H
 
 // Include all C++ headers here:
-#include <cstddef>
-#include <stdexcept>
-#include <iostream>
```
Comment (isuruf) [inline symengine/basic.h]: We should add these as the presence of these headers in others are not guaranteed

File: symengine/eval_mpfr.h [hunk 4]
```diff
@@ -7,8 +7,6 @@
 #ifndef SYMENGINE_EVAL_MPFR_H
 #define SYMENGINE_EVAL_MPFR_H
 
-#include <symengine/symengine_config.h>
```
Comment (isuruf) [inline symengine/eval_mpfr.h]: This is needed

File: symengine/basic.h [hunk 5]
```diff
@@ -8,17 +8,13 @@
 #define SYMENGINE_BASIC_H
 
 // Include all C++ headers here:
-#include <cstddef>
```
Comment (isuruf) [inline symengine/basic.h]: `cstddef` needs to be added before `gmpxx.h` for some gmp versions. Therefore this is needed.

**Comment by isuruf** [comment]
I'm not sure whether this PR is the way to go. I've already found some errors and 1 error about cstddef and gmpxx.h from an old PR by @certik. Doing this would undo any other changes I'm not aware of especially with the header orders.

**Comment by srajangarg** [comment]
@isuruf I don't understand, most of your comments are about `#include <symengine/symengine_config.h>`, but isn't that being implicitly included via `basic.h`

File: symengine/eval_arb.h (hunk 1 above)
Comment (isuruf) [inline symengine/eval_arb.h]: Look at the line below, `symengine_config.h` is needed for the defines inside

Comment (srajangarg) [inline symengine/eval_arb.h]: Maybe we can include the `basic.h` outside the defines? Everywhere

Comment (isuruf) [inline symengine/eval_arb.h]: In this file, you need to include `basic.h` as long as `HAVE_SYMENGINE_ARB` is not defined, so only `symengine_config.h` is enough

File: symengine/basic.h (hunk 3 above)
Comment (srajangarg) [inline symengine/basic.h]: They are guaranteed I think. `dict.h`->`mp_class.h`->`mp_wrapper.h`->`symengine_rcp.h` -> these three headers. (without any if directives)

File: symengine/basic.h (hunk 5 above)
Comment (srajangarg) [inline symengine/basic.h]: Same as above, `cstddef` is ensured to be before `gmpxx.h` (in  `mp_class.h`) as `mp_wrapper.h` is included before it.

File: symengine/basic.h (hunk 2 above)
Comment (srajangarg) [inline symengine/basic.h]: Same, included via `dict.h`->`mp_class.h`->`mp_wrapper.h`->`symengine_rcp.h`

**Comment by certik** [comment]
I think this PR is fine --- if there are bugs that this introduces (which there might be), then we should write tests for them, to catch it the next time. Since all tests pass, I think it's ok.

**Comment by Sumith1896** [comment]
@srajangarg Could you fix the conflicts then? We can merge this

**Comment by Sumith1896** [comment]
Also is there a permanent solution for checking for redundant headers? Integrate with CI like we do clang-format.

**Comment by srajangarg** [comment]
Right, I'll think of a permanent solution too. But I'm not sure it will be that trivial, because the script I wrote did not take care of preprocessing directives at all. Maybe will make another PR for that. Till then, I'll merge this.

**Comment by Sumith1896** [comment]
That works

**Comment by Sumith1896** [comment]
Thanks @srajangarg

**Pull request merged**